File: G-for-g-Practice/Longest Palindrome in a String.java



class Solution {
    private int start, maxLength;
    
    public String longestPalindrome(String s) {
        if ((s==null)|| s.length()<1) return "";

        for (int i=0; i<s.length();i++){
            startMiddle(s,i,i);
            startMiddle(s,i,i+1);
        }
        return s.substring(start, start +maxLength);       
    }
    private void startMiddle(String s, int left, int right){
        while (left>=0 && right <s.length() && s.charAt(left)==s.charAt(right)){
            if(right-left+1>maxLength){
                start= left;
                maxLength=right-left+1;
            }
            left--;
            right++;
        }
    }
}

File: G-for-g-Practice/Longest Palindrome in a String.py
class Solution:
    def __init__ (self):
        self.start=0
        self.max_length=0
    
    def longestPalindrome(self, s: str) -> str:
        if not s or len(s)==1:
            return s
        
        for i in range (len(s)):
            self.startMiddle(s,i,i)
            self.startMiddle(s,i,i+1)
        
        return s[self.start: self.start +self.max_length]

    def startMiddle(self, s:str, left:int, right: int):
        while left>=0 and right <len(s) and s[left]==s[right]:
            if right-left+1 > self.max_length:
                self.start =left
                self.max_length =right-left+1
            left-=1
            right+=1    
